gufe/protocols/protocolunit.py

```diff
@@ -0,0 +1,92 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+
+"""The `ProtocolUnit` class should be subclassed for all units to be used as
+part of a `ProtocolDAG`.
+
+"""
+
+import abc
+from os import PathLike
+from typing import Iterable, List, Dict, Any
+
+from .results import ProtocolUnitResult
+
+
+class ProtocolUnit(abc.ABC):
+    """A unit of work computable by
+
+    """
+
+    def __init__(
+            self,
+            settings: "ProtocolSettings", # type: ignore
+            *dependencies: "ProtocolUnit",
+            **kwargs
+        ):
+
+        self._settings = settings
+        self._dependencies = dependencies
+
+        if self._dependencies:
+            self._status = "WAITING"
+        else:
+            self._status = "READY"
+
+        self._kwargs = kwargs
+
+    @property
+    def settings(self):
+        return self._settings
+
+    def execute(self, block=True) -> ProtocolUnitResult:
+        """Given `ProtocolUnitResult`s from dependencies, execute this `ProtocolUnit`.
+
+        Parameters
+        ----------
+        block : bool
+            If `True`, block until execution completes; otherwise run in its own thread.
+
+        """
+        if block:
+            dep_results = [dep.result for dep in self._dependencies]  # type: ignore
+
+            self._status = "RUNNING"
+            out = self._execute(dep_results)
+            self._status = "COMPLETE"
+            self._result = ProtocolUnitResult(
+                                name=self.__class__.__name__,
+                                dependencies=dep_results, 
+                                **out)
+
+        else:
+            #TODO: wrap in a thread; update status
+            ...
+
+        return self._result
+
+
+    @abc.abstractmethod
+    def _execute(self, dependency_results: List[ProtocolUnitResult]) -> Dict[str, Any]:
+        ...
+
+    @property
+    def result(self) -> ProtocolUnitResult:
+        """Return `ProtocolUnitResult` for this `ProtocolUnit`.
+
+        Requires `status` == "COMPLETE"; exception raised otherwise.
+
+        """
+        return self._result
+
+    @property
+    def status(self):
```

Comment:
I started sketching out what the state diagram for this would look like if this class didn't store the state in an attribute:
- if there's completed output then its FINISHED, either SUCCESS or ERROR
- if there's something in the scheduler for this, it's RUNNING
- if it has unresolved dependencies, it's WAITING
- else it's READY

```mermaid
  graph TD;
      WAITING-->READY;
      READY-->RUNNING;
      RUNNING-->FAILED;
      RUNNING-->COMPLETED;
```

This would require the Unit to see a) the results store to determine if output exists, b) the scheduler to determine if a job of this type exists...

Then I realised that that's not really the job of this object, it's just meant to encode how to run the thing, and all this knowledge of status should be external to this class, so I'd remove `status` on Unit.  Similarly I'd remove the `result()` and `get_artifacts()` methods, these also require knowledge of state (knowing if/where the Unit was ran).  Instead just make a Worker responsible for capturing the output of `execute()` and store the tuple of `(Unit, Result)` somewhere.  So all in all, this is treading on the toes of the scheduler too much and should get stripped down.